README.txt
Fire N' Blood

This mod adds several new mobs into the game!

New Features Coming Soon!


changelog.txt
Build: 1.16.4-35.1.4 - Tue Nov 24 17:14:21 GMT 2020
	csh2001331:
		Fix dimType not encode to registries (#7513)

=========
Build: 1.16.4-35.1.3 - Mon Nov 23 03:45:21 GMT 2020
	richard:
		Add support for referencing forge's resources and specifying existing mods to data generators (#7456)

=========
Build: 1.16.4-35.1.2 - Sun Nov 22 22:01:29 GMT 2020
	cpw:
		Fix #6692
		
		Thanks @AterAnimAvis for help and suggestions
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.16.4-35.1.1 - Sat Nov 21 19:57:45 GMT 2020
	cyborgmas18:
		Fix translucent blocks not rendering properly when moved by pistons on fabulous (#7441)

=========
Build: 1.16.4-35.1.0 - Fri Nov 20 20:54:30 GMT 2020
	lex:
		Bump version for RB.

=========
Build: 1.16.4-35.0.22 - Fri Nov 20 18:24:46 GMT 2020
	35673674+alcatrazEscapee:
		Fix MC-194811 - Removing structures causes chunk save errors.

	diesieben07:
		Fix food bar not rendering when non-living entities are mounted (#7446)

=========
Build: 1.16.4-35.0.20 - Fri Nov 20 18:18:37 GMT 2020
	diesieben07:
		Fix concurrency issue in StartupMessageManager (#7483)

=========
Build: 1.16.4-35.0.19 - Fri Nov 20 18:11:25 GMT 2020
	35673674+alcatrazEscapee:
		Fix forge registry types that have private constructors (BlockStateProviderType, BlockPlacerType, FoliagePlacerType, TreeDecoratorType) (#7482)

=========
Build: 1.16.4-35.0.18 - Sun Nov 15 16:08:55 GMT 2020
	cpw:
		Fix problem where the absence of forge and MC in later loading triggered a weird crash.
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.16.4-35.0.17 - Sat Nov 14 01:41:38 GMT 2020
	cpw:
		Fix NPE problem when modid is mismatched.
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.16.4-35.0.16 - Sat Nov 14 00:18:29 GMT 2020
	cpw:
		Return the processed modlist data to the upstream system, not the raw unprocessed data. Should resolve the problem
		that mixin complains about missing info when it's in a "crashed" situation.
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.16.4-35.0.15 - Wed Nov 11 09:09:45 GMT 2020
	cyborgmas18:
		Fix dims on servers not loading the first run (#7445)

=========
Build: 1.16.4-35.0.14 - Tue Nov 10 20:24:33 GMT 2020
	diesieben07:
		Allow for custom argument types by filtering them on vanilla connections (#7463)

=========
Build: 1.16.4-35.0.13 - Tue Nov 10 20:13:34 GMT 2020
	Bward7864:
		Add Player context to AnvilUpdateEvent (#7442)

=========
Build: 1.16.4-35.0.12 - Tue Nov 10 20:02:33 GMT 2020
	git:
		Don't overwrite PlayerAdvancement's player with a fake one (#7454)
		
		If a FakePlayer is constructed with the same UUID/GameProfile as a
		currently active player, the player object within PlayerAdvancement is
		set to the fake player.
		
		As fake players cannot receive advancements (see grantCriterion), this
		prevents the actual player from receiving any advancements until they
		change dimension/relog.

=========
Build: 1.16.4-35.0.11 - Tue Nov 10 19:46:01 GMT 2020
	curle:
		Add BannerBlock.forColor to SAS to fix banner.toItem() on servers.

=========
Build: 1.16.4-35.0.10 - Tue Nov 10 19:36:40 GMT 2020
	Boy132:
		Call Harvest check event even if the block doesn't require any tool (#7414)

=========
Build: 1.16.4-35.0.9 - Tue Nov 10 19:25:26 GMT 2020
	lex:
		Use linked collections to stabilize order when adding new registry entries.

	xfacthd:
		Render local player when the renderViewEntity is not the local player (#7216)

=========
Build: 1.16.4-35.0.7 - Sun Nov 08 04:12:02 GMT 2020
	cpw:
		Some help in ExtensionPoint regarding DISPLAYTEST. Let's improve those tick rates!
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.16.4-35.0.6 - Sun Nov 08 03:30:37 GMT 2020
	cpw:
		Handle erroring during early mod construction phases, and actually report that into the error UI
		by doing armslength exception handling and propagation.
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.16.4-35.0.5 - Sun Nov 08 01:48:22 GMT 2020
	cpw:
		Support backwards loading 1.16.3 mods in 1.16.4, because we are able to do that. Tweak loading a bit to be smarter about dependency errors of various kinds.
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.16.4-35.0.4 - Sat Nov 07 20:10:44 GMT 2020
	David Quintana:
		Add FluidStack codec, and a test mod that verifies its behaviour matches the existing write/read logic.
		Add missing license headers.

=========
Build: 1.16.4-35.0.3 - Sat Nov 07 02:11:47 GMT 2020
	cpw:
		Update event bus version (should be 3.0.5)
		
		Tweak MDK to automatically populate recommended values for things like forge version, loader version and minecraft version.
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.16.4-35.0.2 - Wed Nov 04 00:07:58 GMT 2020
	diesieben07:
		Workaround for modded attributes on vanilla entities logging warnings on vanilla clients
		
		Switch to a handler-based approach per packet class, simplify patch, add some docs
		
		Add license header and docs
		
		Switch to a network handler based approach
		
		Switch to a network handler based approach
		
		revert build gradle change
		
		Simplify Packet patch
		
		Add license header to IForgePacket
		
		Revert ForgeHooks changes
		
		Less hacky and more generic way to filter packets for vanilla connections
		
		_Actually_ Order SAS.
		It's early
		
		Order SAS
		
		simplify vanilla connection check
		
		Remove unneeded import change

=========
Build: 1.16.4-35.0.1 - Mon Nov 02 19:57:45 GMT 2020
	cpw:
		Fix early sorting bug that meant the file deduping was causing an error, to, well, not cause it, so we can get to an error screen.
		
		Signed-off-by: cpw <cpw+github@weeksfamily.ca>

=========
Build: 1.16.4-35.0.0 - Mon Nov 02 19:15:50 GMT 2020
	lex:
		1.16.4 Initial Update

=========
Build: 1.16.3-34.1.42 - Mon Nov 02 04:30:53 GMT 2020
	lex:
		Fix Biomes not properly copying over ids loaded from the save to the new Dynamic Registry.

=========

src/main/java/com/Polarice3/FireNBlood/FNBConfig.java
package com.Polarice3.FireNBlood;

import com.electronwill.nightconfig.core.file.CommentedFileConfig;
import com.electronwill.nightconfig.core.io.WritingMode;
import net.minecraftforge.common.ForgeConfigSpec;

import java.io.File;

public class FNBConfig {

    public static final ForgeConfigSpec.Builder BUILDER = new ForgeConfigSpec.Builder();
    public static final ForgeConfigSpec SPEC;

    public static final ForgeConfigSpec.ConfigValue<Integer> MaxSouls;
    public static final ForgeConfigSpec.ConfigValue<Integer> VexCost;
    public static final ForgeConfigSpec.ConfigValue<Integer> FangCost;
    public static final ForgeConfigSpec.ConfigValue<Integer> RoarCost;
    public static final ForgeConfigSpec.ConfigValue<Integer> ZombieCost;
    public static final ForgeConfigSpec.ConfigValue<Integer> SkeletonCost;
    public static final ForgeConfigSpec.ConfigValue<Integer> CrippleCost;
    public static final ForgeConfigSpec.ConfigValue<Integer> SpiderlingCost;
    public static final ForgeConfigSpec.ConfigValue<Integer> BrainEaterCost;
    public static final ForgeConfigSpec.ConfigValue<Integer> TeleportCost;
    public static final ForgeConfigSpec.ConfigValue<Integer> SoulSkullCost;

    public static final ForgeConfigSpec.ConfigValue<Integer> VexDuration;
    public static final ForgeConfigSpec.ConfigValue<Integer> FangDuration;
    public static final ForgeConfigSpec.ConfigValue<Integer> RoarDuration;
    public static final ForgeConfigSpec.ConfigValue<Integer> ZombieDuration;
    public static final ForgeConfigSpec.ConfigValue<Integer> SkeletonDuration;
    public static final ForgeConfigSpec.ConfigValue<Integer> CrippleDuration;
    public static final ForgeConfigSpec.ConfigValue<Integer> SpiderlingDuration;
    public static final ForgeConfigSpec.ConfigValue<Integer> BrainEaterDuration;

    public static final ForgeConfigSpec.ConfigValue<Integer> BrainEaterXPCost;

    public static final ForgeConfigSpec.ConfigValue<Integer> VexCooldown;
    public static final ForgeConfigSpec.ConfigValue<Integer> ZombieCooldown;
    public static final ForgeConfigSpec.ConfigValue<Integer> SkeletonCooldown;

    public static final ForgeConfigSpec.ConfigValue<Integer> UndeadSouls;
    public static final ForgeConfigSpec.ConfigValue<Integer> AnthropodSouls;
    public static final ForgeConfigSpec.ConfigValue<Integer> IllagerSouls;
    public static final ForgeConfigSpec.ConfigValue<Integer> VillagerSouls;
    public static final ForgeConfigSpec.ConfigValue<Integer> TaillessSouls;
    public static final ForgeConfigSpec.ConfigValue<Integer> PiglinSouls;
    public static final ForgeConfigSpec.ConfigValue<Integer> EnderDragonSouls;
    public static final ForgeConfigSpec.ConfigValue<Integer> PlayerSouls;
    public static final ForgeConfigSpec.ConfigValue<Integer> DefaultSouls;

    public static final ForgeConfigSpec.ConfigValue<Integer> MaxEnchant;

    public static final ForgeConfigSpec.ConfigValue<Integer> HiredTimer;
    public static final ForgeConfigSpec.ConfigValue<Integer> DyingTimer;

    public static final ForgeConfigSpec.ConfigValue<Integer> CraftingSouls;
    public static final ForgeConfigSpec.ConfigValue<Integer> CultistSpawnFreq;
    public static final ForgeConfigSpec.ConfigValue<Integer> MRabbitMax;

    public static final ForgeConfigSpec.ConfigValue<Float> HealAmount;

    public static final ForgeConfigSpec.ConfigValue<Boolean> HexerSpawn;
    public static final ForgeConfigSpec.ConfigValue<Boolean> CultistsSpawn;

    public static final ForgeConfigSpec.ConfigValue<Boolean> SoulRepair;
    public static final ForgeConfigSpec.ConfigValue<Boolean> TotemUndying;
    public static final ForgeConfigSpec.ConfigValue<Boolean> StarterTotem;

    public static final ForgeConfigSpec.ConfigValue<Boolean> ProfanedTowerGen;
    public static final ForgeConfigSpec.ConfigValue<Boolean> ProfanedShrineGen;
    public static final ForgeConfigSpec.ConfigValue<Boolean> TavernGen;
    public static final ForgeConfigSpec.ConfigValue<Boolean> DarkManorGen;
    public static final ForgeConfigSpec.ConfigValue<Boolean> PortalOutpostGen;
    public static final ForgeConfigSpec.ConfigValue<Boolean> TotemGen;

    public static final ForgeConfigSpec.ConfigValue<Boolean> VizierMinion;

    static {
        BUILDER.push("General");
        MaxSouls = BUILDER.comment("Totem Maximum Soul Count, Default: 10000")
                .defineInRange("maxSouls", 10000, 100, 1000000);
        SoulRepair = BUILDER.comment("Dark and Necro Robes repair themselves using Soul Energy, Default: true")
                .define("soulrepair", true);
        TotemUndying = BUILDER.comment("Totem of Souls will save the Player if full of Soul Energy, Default: true")
                .define("totemundying", true);
        MaxEnchant = BUILDER.comment("Soul Eater Maximum Enchantment Level, Default: 3")
                .defineInRange("maxEnchant", 3, 1, 10);
        StarterTotem = BUILDER.comment("Gives Players a Totem of Souls when first entering World, Default: false")
                .define("startertotem", false);
        CraftingSouls = BUILDER.comment("How much Souls is consumed when crafting with Totem, Default: 1")
                .defineInRange("craftsouls", 1, 0, 1000000);
        BUILDER.pop();
        BUILDER.push("Protectors");
        HiredTimer = BUILDER.comment("Hired Timer, Default: 24000")
                .defineInRange("hiredtimer", 24000, 0, 72000);
        DyingTimer = BUILDER.comment("Dying Timer, Default: 12000")
                .defineInRange("dyingtimer", 12000, 0, 72000);
        HealAmount = BUILDER.comment("Heal Amount after being fed (Always put an F at the end of the value!), Default: 10.0F")
                .define("healamount", 10.0F);
        BUILDER.pop();
        BUILDER.push("Soul Taken");
        UndeadSouls = BUILDER.comment("Undead Killed, Default: 5")
                .defineInRange("undeadsouls", 5, 0, 1000000);
        AnthropodSouls = BUILDER.comment("Anthropods Killed, Default: 5")
                .defineInRange("anthropodsouls", 5, 0, 1000000);
        IllagerSouls = BUILDER.comment("Illagers, Witches, Cultists, Protectors Killed, Default: 25")
                .defineInRange("illagersouls", 25, 0, 1000000);
        VillagerSouls = BUILDER.comment("Villagers Killed, Default: 50")
                .defineInRange("villagersouls", 50, 0, 1000000);
        TaillessSouls = BUILDER.comment("Tailless Killed, Default: 45")
                .defineInRange("taillesssouls", 45, 0, 1000000);
        PiglinSouls = BUILDER.comment("Non-Undead Piglin Killed, Default: 10")
                .defineInRange("piglinsouls", 10, 0, 1000000);
        EnderDragonSouls = BUILDER.comment("Ender Dragon Killed, Default: 1000")
                .defineInRange("enderdragonsouls", 1000, 0, 1000000);
        PlayerSouls = BUILDER.comment("Players Killed, Default: 100")
                .defineInRange("playersouls", 100, 0, 1000000);
        DefaultSouls = BUILDER.comment("Others Killed, Default: 5")
                .defineInRange("othersouls", 5, 0, 1000000);
        BUILDER.pop();
        BUILDER.push("Spell Costs");
        VexCost = BUILDER.comment("Vex Spell Cost, Default: 18")
                .defineInRange("vexcost", 18, 0, 1000000);
        FangCost = BUILDER.comment("Fang Spell Cost, Default: 8")
                .defineInRange("fangcost", 8, 0, 1000000);
        RoarCost = BUILDER.comment("Roaring Spell Cost, Default: 10")
                .defineInRange("bitecost", 10, 0, 1000000);
        ZombieCost = BUILDER.comment("Necroturgy Spell Cost, Default: 5")
                .defineInRange("zombiecost", 5, 0, 1000000);
        SkeletonCost = BUILDER.comment("Osseous Spell Cost, Default: 8")
                .defineInRange("skeletoncost", 8, 0, 1000000);
        CrippleCost = BUILDER.comment("Crippling Spell Cost, Default: 15")
                .defineInRange("cripplecost", 15, 0, 1000000);
        SpiderlingCost = BUILDER.comment("Spiderling Spell Cost per second, Default: 2")
                .defineInRange("spiderlingcost", 2, 0, 1000000);
        BrainEaterCost = BUILDER.comment("Brain Eater Spell Cost per second, Default: 5")
                .defineInRange("braincost", 5, 0, 1000000);
        BrainEaterXPCost = BUILDER.comment("How much Experience the above spell Cost per heal, Default: 10")
                .defineInRange("brainxpcost", 10, 0, 1000);
        TeleportCost = BUILDER.comment("Teleport Spell Cost, Default: 5")
                .defineInRange("teleportcost", 5, 0, 1000000);
        SoulSkullCost = BUILDER.comment("Soul Skull Spell Cost, Default: 25")
                .defineInRange("soulskullcost", 25, 0, 1000000);
        BUILDER.pop();
        BUILDER.push("Casting Time");
        VexDuration = BUILDER.comment("Time to cast Vex Spell, Default: 100")
                .defineInRange("vextime", 100, 0, 72000);
        FangDuration = BUILDER.comment("Time to cast Fang Spell, Default: 80")
                .defineInRange("fangtime", 80, 0, 72000);
        RoarDuration = BUILDER.comment("Time to cast Roaring Spell, Default: 40")
                .defineInRange("roartime", 40, 0, 72000);
        ZombieDuration = BUILDER.comment("Time to cast Necroturgy Spell, Default: 60")
                .defineInRange("zombietime", 60, 0, 72000);
        SkeletonDuration = BUILDER.comment("Time to cast Osseous Spell, Default: 60")
                .defineInRange("skeletontime", 60, 0, 72000);
        CrippleDuration = BUILDER.comment("Time to cast Crippling Spell, Default: 20")
                .defineInRange("crippletime", 20, 0, 72000);
        SpiderlingDuration = BUILDER.comment("Time to cast Spiderling Spell per second, Default: 20")
                .defineInRange("spiderlingtime", 20, 0, 72000);
        BrainEaterDuration = BUILDER.comment("Time to cast Brain Eater Spell per second, Default: 20")
                .defineInRange("braineatertime", 20, 0, 72000);
        BUILDER.pop();
        BUILDER.push("Summon Down Duration");
        VexCooldown = BUILDER.comment("Vex Spell Cooldown, Default: 340")
                .defineInRange("vexcooldown", 340, 0, 72000);
        ZombieCooldown = BUILDER.comment("Necroturgy Spell Cooldown, Default: 250")
                .defineInRange("zombiecooldown", 250, 0, 72000);
        SkeletonCooldown = BUILDER.comment("Osseous Spell Cooldown, Default: 280")
                .defineInRange("skeletoncooldown", 280, 0, 72000);
        BUILDER.pop();
        BUILDER.push("Mob Spawning");
        HexerSpawn = BUILDER.comment("Hexer Spawning, Default: true")
                .define("hexerspawn", true);
        CultistsSpawn = BUILDER.comment("Cultists Spawning, Default: true")
                .define("cultistsspawn", true);
        CultistSpawnFreq = BUILDER.comment("Spawn Frequency for Cultists, Default: 12000")
                .defineInRange("cultistsspawnfreq", 12000, 0, 72000);
        MRabbitMax = BUILDER.comment("Maximum amount of Mutant Rabbits can spawn, Default: 16")
                .defineInRange("mrabbitmax", 16, 0, 100);
        BUILDER.pop();
        BUILDER.push("Structure Generation");
        TavernGen = BUILDER.comment("Tavern Generates in the World, Default: true")
                .define("taverngen", true);
        ProfanedTowerGen = BUILDER.comment("Profaned Tower Generates in the World, Default: true")
                .define("profanedtowergen", true);
        ProfanedShrineGen = BUILDER.comment("Profaned Shrine Generates in the World, Default: true")
                .define("profanedshrinegen", true);
        DarkManorGen = BUILDER.comment("Dark Manor Generates in the World, Default: true")
                .define("darkmanorgen", true);
        PortalOutpostGen = BUILDER.comment("Portal Outpost Generates in the World, Default: true")
                .define("portaloutpostgen", true);
        TotemGen = BUILDER.comment("Totems Generates in the World, Default: true")
                .define("totemgen", true);
        BUILDER.pop();
        BUILDER.push("Misc");
        VizierMinion = BUILDER.comment("Viziers spawn Vexes instead of Irks, Default: false")
                .define("vizierminion", false);
        BUILDER.pop();
        SPEC = BUILDER.build();
    }

    public static void loadConfig(ForgeConfigSpec config, String path) {
        final CommentedFileConfig file = CommentedFileConfig.builder(new File(path)).preserveInsertionOrder().sync().autosave().writingMode(WritingMode.REPLACE).build();
        file.load();
        config.setConfig(file);
    }

}
